lightning-invoice/src/payment.rs

```diff
@@ -2424,4 +2424,143 @@ mod tests {
 		nodes[0].node.process_pending_events(&invoice_payer);
 		assert!(expected_events.borrow().is_empty());
 	}
+
+	///	Helper function to forward HTLC update and RevokeAndAck messages
+	fn handle_message_send_events(sender:&Node, node_indices_by_id: &HashMap<PublicKey, usize>, nodes: &Vec<Node>) {
+		let mut to_notify : Vec<&Node>=Vec::new();
+		for msg in sender.node.get_and_clear_pending_msg_events() {
+			match msg {
+				MessageSendEvent::SendRevokeAndACK { node_id, msg } => {
+					println!("Sendrevoke {:?}", msg);
+					let idx = node_indices_by_id[&node_id];
+					let other_node=&nodes[idx];
+					other_node.node.handle_revoke_and_ack(&sender.node.node_id(), &msg);
+					to_notify.push(other_node);
+					// handle_message_send_events(&other_node, node_indices_by_id, nodes);
+				}
+				MessageSendEvent::UpdateHTLCs { node_id, updates } => {
+					println!("Update {:?}", updates);
+					let idx = node_indices_by_id[&node_id];
+					let other_node=&nodes[idx];
+					for add_htlc in updates.update_add_htlcs {
+						other_node.node.handle_update_add_htlc(&sender.node.node_id(), &add_htlc);
+					}
+					for update_fail_htlc in updates.update_fail_htlcs {
+						other_node.node.handle_update_fail_htlc(&sender.node.node_id(), &update_fail_htlc);
+					}
+					other_node.node.handle_commitment_signed(&sender.node.node_id(), &updates.commitment_signed);
+					to_notify.push(other_node);
+
+					// handle_message_send_events(&other_node, node_indices_by_id, nodes);
+				},
+				_ => panic!("Unexpected event: {:?}", msg),
+			}
+		}
+
+		for ton in to_notify {
+			handle_message_send_events(ton, node_indices_by_id, nodes);
+		}
+		// clear monitors
+		let mut added_monitors = sender.chain_monitor.added_monitors.lock().unwrap();
+		added_monitors.clear();
+	}
+
+	#[test]
+	/// A simpler version of no_extra_retries_on_back_to_back_fail test that demonstrates
+	/// automatically handling and forwarding message send events during a payment inside the test case.
+	fn no_extra_retries_on_back_to_back_fail_automatically_handle_message_send_events() {
```

Comment:
If the coverage will be the same, it makes sense to entirely replace `no_extra_retries_on_back_to_back_fail` with this refactored function. Would be nice to keep the documenting comments there that explain the reasoning behind the test case.